src/Aspire.Hosting.AWS/CDK/CDKExtensions.cs

```diff
@@ -0,0 +1,227 @@
+    public static IResourceBuilder<T> WithReference<T>(this IResourceBuilder<T> builder, IAWSSDKConfig awsSdkConfig)
+        where T : IStackResource
+    {
+        builder.Resource.AWSSDKConfig = awsSdkConfig;
+        return builder;
+    }
+
+    /// <summary>
+    /// Add a reference of a AWS CDK stack to a project. The output parameters of the CloudFormation stack are added to the project IConfiguration.
+    /// </summary>
+    /// <typeparam name="TDestination"></typeparam>
+    /// <param name="builder">The resource builder.</param>
+    /// <param name="stack">The stack resource.</param>
+    /// <param name="configSection">The config section in IConfiguration to add the output parameters.</param>
+    /// <returns></returns>
+    public static IResourceBuilder<TDestination> WithReference<TDestination>(this IResourceBuilder<TDestination> builder, IResourceBuilder<IStackResource> stack, string configSection = Constants.DefaultConfigSection)
+        where TDestination : IResourceWithEnvironment
+    {
+        stack.WithAnnotation(new CloudFormationReferenceAnnotation(builder.Resource.Name));
+
+        builder.WithEnvironment(async ctx =>
+        {
+            // Skip when in publish mode
+            if (ctx.ExecutionContext.IsPublishMode)
+            {
+                return;
+            }
+
+            // When the stack has a AWS credentials profile attached, apply that to the resource
+            if (stack.Resource.AWSSDKConfig != null)
+            {
+                SdkUtilities.ApplySDKConfig(ctx, stack.Resource.AWSSDKConfig, false);
+            }
+
+            // Wait for the stack to be ready
+            if (stack.Resource.ProvisioningTaskCompletionSource is not null)
+            {
+                ctx.Logger?.LogInformation("Waiting on Stack resource {Name} ...", stack.Resource.Name);
+                await stack.Resource.ProvisioningTaskCompletionSource.Task.WaitAsync(ctx.CancellationToken).ConfigureAwait(false);
+            }
+
+            // Get the stack outputs and skip when there are none
+            var stackOutputs = stack.Resource.Outputs;
+            if (stackOutputs == null)
+            {
+                return;
+            }
+
+            configSection = configSection.ToEnvironmentVariables();
+
+            // Add the stack outputs for each child construct to the project configuration
+            var processedOutputs = new List<Output>();
+            foreach (var construct in stack.Resource.ListChildren(builder.ApplicationBuilder.Resources.OfType<IConstructResource>()))
+            {
+                var constructId = construct.Construct.StackUniqueId();
+
+                // Filter the outputs attached to the construct
+                var outputs = stackOutputs.Where(o => o.OutputKey.StartsWith(constructId)).ToList();
+
+                // Compose the output name with the name of the constructs
+                var prefix = configSection.ToEnvironmentVariables();
```

Comment:
This logic didn't work for me when I changes the playground to have the following output parameter.

```
table.AddConstruct<CfnOutput>("TableNameReference", (x => new CfnOutput(x, "TableNameReference", new CfnOutputProps
{
    Value = table.Resource.Construct.TableName
})));
```

I got the psuedo random output name added from the fallback foreach loop after this block.

Reply:
I'll check this, and write some tests around this part. This logic tent to be a bit more complex due to the CfnOutput lookups.

Reply:
What happens here js CDK specific. Before the Synth it set some placeholder values to the strings. To this means we can only access this after the build.

Reply:
Are you saying there is no way to make this work to have a predicable user specified output key? If the only thing we can put in here is the pseudo random output keys when referencing a CDK stack then we should probably have APIs for doing `WithReference` with constructs instead of the stack.